File: KMean.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Random;
import java.util.Scanner;
public class KMean {
	private static double diss,sim;
	private static class K_mean{
		double data[][];
		int cluster[][];
		double centroid[][];
		double L[];
		double U[];
		int attributes,total_clusters,total_data;
		public void display_data() {
			for(int i=0;i<total_data;i++) {
				System.out.print("\nLine "+i+" : ");
				for(int j=0;j<attributes;j++) {
					System.out.format(" .3f ",data[i][j]);
				}
			}
			
		}
		public void display_clusters() {
			for(int i=0;i<total_clusters;i++) {
				System.out.print("\nCluster "+i+"= "+cluster[i][0]+" :");
				for(int j=1;j<cluster[i][0];j++) {
					System.out.print(cluster[i][j]+" ");
				}
			}
		}
		public void display_centroid() {
			for(int i=0;i<total_clusters;i++) {
				System.out.print("\nCentroid "+i+" : ");
				for(int j=0;j<attributes;j++) {
					System.out.format("%.3f ",centroid[i][j]);
				}
			}
		}
		public double eu_distance(double v1[],double v2[]) {
			double sum=0;
			for(int i=0;i<v1.length;i++) {
				sum+=Math.pow(v1[i]-v2[i], 2);
			}
			return Math.sqrt(sum);
		}
		public void create_cluster() {
			diss=sim=0;
			for(int i=0;i<total_clusters;i++) {
				cluster[i][0]=0;
			}
			for(int i=0;i<total_data;i++) {
				double min=eu_distance(centroid[0],data[i]);
				double max=min;
				int index=0;
				for(int j=0;j<total_clusters;j++) {
					double d=eu_distance(centroid[j],data[i]);
					if(d<min) {
						min=d;
						index=j;
					}
					if(d>max) {
						max=d;
					}
				}
				cluster[index][0]++;
				cluster[index][cluster[index][0]]=i;
				sim+=min;
				diss+=max;
			}
		}
		public void update_centroid() {
			for(int i=0;i<total_clusters;i++) {
				for(int j=0;j<attributes;j++) {
					double sum=0;
					if(cluster[i][0]!=0) {
						for(int k=1;k<cluster[i][0];k++) {
							sum+=data[cluster[i][k]][j];
						}
					}
					centroid[i][j]=sum/cluster[i][0];
				}
			}
		}
		public void create_centroid() {
			Random r=new Random();
			for(int i=0;i<total_clusters;i++) {
				for(int j=0;j<attributes;j++) {
					centroid[i][j]=L[j]+(U[j]-L[j])*r.nextDouble();
				}
			}
		}
		public void read_data() {
			try {
				Scanner sc=new Scanner(new File("G:\\biodata.csv"));
				String[] Line=new String[3];
	            Line=sc.nextLine().split(",");
	            total_data=Integer.parseUnsignedInt(Line[0]);
	            attributes=Integer.parseUnsignedInt(Line[1]);
	            total_clusters=Integer.parseUnsignedInt(Line[2]);
	            
	            data=new double[total_data][attributes];
	            cluster=new int[total_clusters][total_data+1];
	            centroid=new double[total_clusters][attributes];
	            L=new double[attributes];
	            U=new double[attributes];
	            
	            String[] data_line=new String[attributes+1];// Including actual cluster number in 1st column
	       	 	data_line=sc.nextLine().split(",");
		       	 for(int j=1;j<(attributes+1);j++){
	         		double val;
	         		val=Double.parseDouble(data_line[j-1]);
	         		data[0][j-1]=val;
	         		L[j-1]=val;
	         		U[j-1]=val;
	         	}
		       int 	index=1;
	            while(index<total_data)
	            {
	            	data_line=sc.nextLine().split(",");
		            for(int j=1;j<(attributes);j++)
		            {
		            	double Val;
		            	Val=Double.parseDouble(data_line[j]);
		            	data[index][j-1]=Val;
		            	if(L[j-1]>Val)L[j-1]=Val;
		            	if(U[j-1]<Val)U[j-1]=Val;
		            }
		            index++;
	            }       
			} catch (FileNotFoundException e) {
				e.printStackTrace();
				System.exit(0);
			}	
		}
		public void execute() {
			double tempD,tempS;
			System.out.println("\n===========Initial Centroids and Clusters==========\n");
			create_centroid();
			display_centroid();
			create_cluster();
			display_clusters();
			System.out.format("\nSimilairty: %.3f \tDisimilarity: %.3f",sim,diss);
			tempD=diss;
			tempS=sim;
			System.out.println("\n\n===========Final Centroids and Clusters==========\n");
			for(int i=0;i<99;i++) {
//				System.out.println("\nIteration "+i+"\n");
				update_centroid();
				create_cluster();
				if(tempD==diss && tempS==sim) {
					System.out.print("\nIteration: "+i);
					break;
				}
				tempD=diss;
				tempS=sim;
//				display_clusters();
			}
			display_centroid();
			display_clusters();
			System.out.format("\nSimilairty: %.3f \tDisimilarity: %.3f",sim,diss);
		}
	}
	public static void main(String[] args) {
		K_mean k=new K_mean();
		k.read_data();
		k.execute();
	}
}

File: KMean.cpp
#include<iostream>
#include<iomanip>
#include<stdlib.h>
#include<math.h>
#include<fstream>
#include<vector>
using namespace std;
class KMean{
	vector<vector<double> >data; 
	vector<vector<double> >centroid; 
	vector<vector<int> >cluster;
	vector<double>L;
	vector<double>U;
	int attributes,total_clusters,total_data;
	double similar;
	double dissimilar;
public:
	void display_clusters();
	void display_centroids();
	void display_data();
	void get_data(); //read data from file and resize corresponding data,centroid and cluster space
	void execute(); //create initial centroids and start iterating until optimal clusters found
	void create_cluster();
	void create_centroid(); //create initial centroids randomly
	void update_centroid();
	double manhattan_distance(vector<double>,vector<double>); 
	double eu_distance(vector<double>,vector<double>);
};
void KMean::display_data(){
	for(int i=0;i<total_data;i++){
		cout<<"\nLine "<<i<<": ";
		for(int j=0;j<attributes;j++){
			cout<<data[i][j]<<" ";
		}
	}
}
void KMean::display_clusters(){
	for(int i=0;i<total_clusters;i++){
		cout<<"\nCluster "<<i<<" ["<<cluster[i].size()<<"]: ";
		for(int j=0;j<cluster[i].size();j++){
			cout<<cluster[i][j]<<" ";
		}
	}
}
void KMean::display_centroids(){
	for(int i=0;i<total_clusters;i++){
		cout<<"\nCentroid "<<i<<" : ";
		for(int j=0;j<attributes;j++){
			cout<<setprecision(2)<<centroid[i][j]<<" ";
		}
	}
}
double KMean::manhattan_distance(vector<double>v1,vector<double>v2){
	double sum=0;
	for(int i=0;i<v1.size();i++){
		sum+=abs(v1[i]-v2[i]);
	}
	return sum;
}
double KMean::eu_distance(vector<double>v1,vector<double>v2){
	double sum=0;
	for(int i=0;i<v1.size();i++){
		sum+=pow((v1[i]-v2[i]),2);
	}
	return pow(sum,1/2.);
}
void KMean::create_cluster(){
	double similarity=0;
	double dis=0;
	for(int i=0;i<total_clusters;i++){
		cluster[i].resize(0);		//for all clusters resize their size to zero hence eventually emptying them
	}
	for(int i=0;i<total_data;i++){
		int index=0;
		double min=eu_distance(centroid[0],data[i]);
		double max=min;
		for(int j=0;j<total_clusters;j++){
			double d=eu_distance(centroid[j],data[i]);
			if(d<min){
				min=d;
				index=j;
			}
			if(d>max){
				d=max;
			}
		}
		similarity+=min;
		dis+=max;
		cluster[index].push_back(i); //push the index at rear of the cluster number "index" hence expanding its size as new data added
	}
	similar=similarity;
	dissimilar=dis;
//	cout<<"\n===================\nSimilarityMeasure: "<<setprecision(8)<<similarity<<endl;
	//cout<<"DismilarityMeasure: "<<setprecision(8)<<disimilarity<<endl;
}
void KMean::update_centroid(){
	for(int i=0;i<total_clusters;i++){
		for(int j=0;j<attributes;j++){
			if(cluster[i].size()!=0){ //if cluster i is not empty then update corresponding centroid
				double sum=0;
				for(int k=0;k<cluster[i].size();k++){ //iterate till the end of current cluster
					sum+=data[k][j];
				}
				centroid[i][j]=sum/cluster[i].size();	//find out average for corresponding cluster X corresponding attribute and insert into centroid
			}
		}
	}
}
void KMean::create_centroid(){
	for(int i=0;i<total_clusters;i++){
		for(int j=0;j<attributes;j++){
			centroid[i][j]=L[j]+(U[j]-L[j])*( ((double) rand() / (RAND_MAX)));
		}
	}
}
void KMean::get_data(){
	ifstream ptr("G:\\biodata.csv");
	string s;
	getline(ptr,s);
	int count=0;
	int index=-1;
	vector<int>v;
	while(count!=3){
		string element;
		index++;
		while(s[index]!=','){
			element+=s[index++];
		}
		v.push_back(atoi(element.c_str()));
		count++;
	}
	total_data=v[0];
	attributes=v[1];
	total_clusters=v[2];
	//total_clusters=2;
	data.resize(total_data,vector<double>(attributes,0)); //resize data matrix into "total_data X attributes" size
	centroid.resize(total_clusters,vector<double>(attributes,0)); //resize the centroid into "total_cluster X attributes"  size
	cluster.resize(total_clusters); //resize the cluster into "total_cluster" size i.e., only rows are initialized, the columns i.e., data can be added using "push_back"
	for(int i=0;i<total_data;i++){
   		getline(ptr,s);
		 count=0;
		 index=-1;
		 while(count!=attributes){
		 	string temp;
		 	index++;
		 	while(s[index]!=','){
		 		temp+=s[index++];
			 }
			 data[i][count]=atof(temp.c_str());
			 count++;
		 }
	}
	for(int i=0;i<attributes;i++){
		L.push_back(data[0][i]);
		U.push_back(data[0][i]);
	}
	for(int j=0;j<attributes;j++){
		for(int i=0;i<total_data;i++){
			if(L[j]>data[i][j]){
				L[j]=data[i][j];
			}
			if(U[j]<data[i][j]){
				U[j]=data[i][j];
			}
		}
	}
	//display_data();	//display inserted data set
}
void KMean::execute(){
	cout<<"\n====================Initial centroids and Clusters=============\n\n";
	create_centroid();
	display_centroids();
	create_cluster();
	display_clusters();
	cout<<"\nSimilarity: "<<fixed<<setprecision(4)<<similar<<"\t Dissimilarity: "<<setprecision(4)<<dissimilar<<endl;
	double tempS,tempD;
	tempS=similar;
	tempD=dissimilar;
	cout<<"\n\n====================Final centroids and Clusters=============\n\n";
	for(int i=0;i<100;i++){
		update_centroid();
		create_cluster();
		if(tempS==similar || tempD==dissimilar){
			cout<<"\nIteration : "<<i<<endl;
			break;
		}
		tempS=similar;
		tempD=dissimilar;
	//	cout<<endl<<"Iteration: "<<i+1<<endl; //for each iteration print the centroids and clusters
	//	display_centroids();
	//	display_clusters();
	}
	display_centroids();
	display_clusters();
	cout<<"\nSimilarity: "<<fixed<<setprecision(4)<<similar<<"\t Dissimilarity: "<<setprecision(4)<<dissimilar<<endl;
}
int main(){
	KMean K;
	K.get_data();
	K.execute();
	return 0;
}
